### PR TITLE
Tighten the definition of CL_IMAGE_{ROW,SLICE}_PITCH queries

### LINKED ISSUE #763: Value returned for CL_IMAGE_ROW_PITCH and CL_IMAGE_SLICE_PITCH queries
Both queries are currently defined rather loosely and the value returned by implementations is not validated by the CTS.

Taking the example of `CL_IMAGE_ROW_PITCH`, it is currently defined as:

> Return calculated row pitch in bytes of a row of elements of the image object given by image.

What "calculated" means isn't made explicit in the description for the query itself but several other sections of the specification describe that pitch "calculation" as "element size * width" which is indeed what most implementations seem to return for the query.

There are cases where a user can observe a linearly tiled representation of the image with a pitch that's different from this "calculated" pitch:
- When an image is created with a host pointer and `image_row_pitch` is set by the application to a non-zero value that is different from the calculated pitch
- When an image is created from a buffer and `image_row_pitch` is set by the application to a non-zero value that is different from the calculated pitch
- When an image that was not created with a host pointer is mapped for host access (the pitch is chosen and returned by the implementation).

This means an application that queries `CL_IMAGE_ROW_PITCH` may get a value back that has no relationship to any observable layout of the image data which makes that query essentially useless. The lack of CTS testing means that returning any value has been considered "conformant" so far.

I suggest we redefine those queries such that they return the pitches used for all application visible linearly tiled representations of the full image's data and update the CTS to require implementations adhere to the new behaviour:
- The data pointed to by an image's host pointer
- The data in a buffer the image was created from
- The mapped view of the image (when the entire image is mapped)

### PR DESCRIPTION
**Pull request by kpet**
Fixes #763

Signed-off-by: Kevin Petit <kevin.petit@arm.com>
Change-Id: I50875d6aa1ffdf6152832c68ac6cf6bdcb434304

### CODE REVIEW

File: api/opencl_runtime_layer.asciidoc [hunk 1]
```diff
@@ -3430,15 +3430,25 @@ include::{generated}/api/version-notes/CL_IMAGE_ELEMENT_SIZE.asciidoc[]
 
 include::{generated}/api/version-notes/CL_IMAGE_ROW_PITCH.asciidoc[]
   | {size_t_TYPE}
-      | Return calculated row pitch in bytes of a row of elements of the
-        image object given by _image_.
+      | Returns the row pictch in bytes of a row of elements of the
```
Comment (bashbaug) [inline api/opencl_runtime_layer.asciidoc:3433]:
```suggestion
      | Returns the row pitch in bytes of a row of elements of the
```

**Review by bashbaug — APPROVED** [review]

File: api/opencl_runtime_layer.asciidoc (hunk 1 above)
Comment (kpet) [reply api/opencl_runtime_layer.asciidoc]: Thanks!

**Comment by bcalidas** [comment]
I feel that it may be premature to make this edit since the concept of linear tiling has not been explained in the spec. Also, since  row_pitch = image_width × size of element applies only when the image is created with host_ptr != NULL, I feel it does not make sense to return this value back.  My suggestion would be -  if the image has linear tiling then image_row_pitch should reflect the true image row pitch as laid out in device memory. In all other cases the queried  image-row_pitch would not have meaning.

**Comment by kpet** [comment]
@bcalidas As discussed in the teleconference I've removed the mention of linear tiling. I agree we haven't introduced the concept to the specification yet and it is not necessary to mention as part of this change. PTAL.

**Comment by bashbaug** [comment]
Discussed on the call May 3rd.  We're not sure if it is possible to create an image from a buffer and to also pass a non-NULL host_ptr.  Need to create an issue for tracking.  Otherwise, ok to merge.

**Comment by bcalidas** [comment]
The current text looks ok.

**Review by alycm — APPROVED** [review]
LGTM.

**Comment by bashbaug** [comment]
I filed https://github.com/KhronosGroup/OpenCL-Docs/issues/789 to track the possible issue we discussed on the call.  Merging!

**Pull request merged**